src/browser/suggestion-box.js

```diff
@@ -190,18 +193,94 @@ define((require, exports, module) => {
   // IO.
 
   let xhrSearch;
-  let xhrHistory;
 
 
   Suggestions.reset = suggestions => {
     if (xhrSearch) xhrSearch.abort();
-    if (xhrHistory) xhrHistory.abort();
     return suggestions.clear();
   }
 
+  const dbHistory = new History().stores.pages
+
+  // Calculates the score for use in suggestions from
+  // a result array `match` of `RegExp#exec`.
+  const score = (pattern, input='', base=0.3, length=0.25) => {
+      const index = 1 - base - length
+      const text = String(input);
+      const count = text.length;
+      const match = pattern.exec(text);
+
+      return !match ? -1 :
+              base +
+              length * Math.sqrt(match[0].length / count) +
+              index * (1 - match.index / count);
+  }
+
+  const Pattern = (input, flags="i") => {
+    try {
+      return RegExp(input, flags)
+    } catch (error) {
+      if (error instanceof SyntaxError) {
+        return RegExp(pattern.escape(input), flags)
+      }
+      throw error
+    }
+  }
+  Pattern.escape = input => input.replace(/[\.\?\*\+\^\$\|\(\)\{\[\]\\]/g, '\\$&')
+
+  const historyService = input => spawn(function*() {
+    const {rows} = yield dbHistory.allDocs({include_docs: true});
+    // Build a query patter from all words and individual words, note that
+    // scoring will take into consideration the length of the match so if we match
+    // multiple words that gets larger score then if we matched just one.
+    const query = Pattern(input.split(/\s+/g).join('[\\s\\S]+') +
+                          '|' + input.split(/\s+/g).join('|'));
+    return rows.map(row => row.doc)
+               .filter(row => row.type === 'Page' && row.title)
+               .map(page => {
+                  // frequency score is ranked from 0-1 not based on quality of
+                  // match but solely on how often this page has been visited in the
+                  // past.
+                  const frequencyScore = 1 - (0.7 / (1 + page.visits.length));
+                  // Title and uri are scored based of input length & match length
+                  // and match index.
+                  const titleScore = score(query, page.title);
+                  const uriScore = score(query, page.uri);
+
+                  // Store each score just for debuging purposes.
+                  page.frequencyScore = frequencyScore;
+                  page.titleScore = titleScore;
+                  page.uriScore = uriScore;
+
+                  // Total score is ranked form `-1` to `1`. Score is devided into
```

Comment:
> Idea: would math be easier if score was 0..1 rather than -1..1?

Negative values are there so we that score of individual field can actually have a negative impact of overal score. Probably it still could be made to be in 0 to 1 range but seems less obvious to me.